File: bindings/r/R/safe_currency.R
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

#' Safe currency operations with proper decimal handling.
#'
#' Provides currency code validation, Money S3 class with arithmetic operations,
#' and safe formatting for financial calculations.

#' Named vector of ISO 4217 currency codes with their decimal places.
#'
#' @export
currency_codes <- c(
  # Major currencies
  USD = 2L, EUR = 2L, GBP = 2L, JPY = 0L, CHF = 2L, AUD = 2L, CAD = 2L,
  NZD = 2L, CNY = 2L, HKD = 2L, SGD = 2L, SEK = 2L, NOK = 2L, DKK = 2L,
  # Additional currencies
  MXN = 2L, BRL = 2L, INR = 2L, RUB = 2L, ZAR = 2L, KRW = 0L, TWD = 2L,
  THB = 2L, MYR = 2L, IDR = 2L, PHP = 2L, VND = 0L, PLN = 2L, CZK = 2L,
  HUF = 2L, TRY = 2L, ILS = 2L, AED = 2L, SAR = 2L, EGP = 2L, NGN = 2L,
  KES = 2L, GHS = 2L, PKR = 2L, BDT = 2L, CLP = 0L, COP = 2L, PEN = 2L,
  ARS = 2L, VES = 2L, UAH = 2L, RON = 2L, BGN = 2L, HRK = 2L, ISK = 0L,
  # Zero-decimal currencies
  BIF = 0L, CLP = 0L, DJF = 0L, GNF = 0L, ISK = 0L, JPY = 0L, KMF = 0L,

KRW = 0L, PYG = 0L, RWF = 0L, UGX = 0L, VND = 0L, VUV = 0L, XAF = 0L,
  XOF = 0L, XPF = 0L,
  # Three-decimal currencies
  BHD = 3L, IQD = 3L, JOD = 3L, KWD = 3L, LYD = 3L, OMR = 3L, TND = 3L,
  # Precious metals and special
  XAU = 6L, XAG = 6L, XPT = 6L, XPD = 6L, BTC = 8L, ETH = 18L
)

#' Get decimal places for a currency code.
#'
#' @param currency_code ISO 4217 currency code
#' @return Number of decimal places or NA if unknown
#' @export
currency_decimals <- function(currency_code) {
  if (is.na(currency_code) || is.null(currency_code)) {
    return(NA_integer_)
  }

  code <- toupper(currency_code)
  if (code %in% names(currency_codes)) {
    currency_codes[[code]]
  } else {
    NA_integer_
  }
}

#' Check if currency code is valid.
#'
#' @param currency_code Currency code to check
#' @return TRUE if valid ISO 4217 code
#' @export
is_valid_currency <- function(currency_code) {
  if (is.na(currency_code) || is.null(currency_code)) {
    return(FALSE)
  }

  toupper(currency_code) %in% names(currency_codes)
}

#' Create a Money object.
#'
#' Money is stored internally as the smallest currency unit (e.g., cents for USD).
#' This avoids floating-point precision issues.
#'
#' @param amount Amount in major units (e.g., dollars)
#' @param currency ISO 4217 currency code
#' @return Money S3 object or NULL on invalid input
#' @export
money <- function(amount, currency) {
  if (is.na(amount) || is.null(amount)) {
    return(NULL)
  }

  currency <- toupper(currency)
  decimals <- currency_decimals(currency)

  if (is.na(decimals)) {
    return(NULL)
  }

  # Convert to minor units (cents, etc.)
  minor_units <- round(amount * (10 ^ decimals))

  structure(
    list(
      minor_units = as.integer(minor_units),
      currency = currency,
      decimals = decimals
    ),
    class = "Money"
  )
}

#' Create Money from minor units.
#'
#' @param minor_units Amount in smallest currency unit
#' @param currency ISO 4217 currency code
#' @return Money S3 object or NULL on invalid input
#' @export
money_from_minor <- function(minor_units, currency) {
  if (is.na(minor_units) || is.null(minor_units)) {
    return(NULL)
  }

  currency <- toupper(currency)
  decimals <- currency_decimals(currency)

  if (is.na(decimals)) {
    return(NULL)
  }

  structure(
    list(
      minor_units = as.integer(minor_units),
      currency = currency,
      decimals = decimals
    ),
    class = "Money"
  )
}

#' Check if object is Money.
#'
#' @param x Object to check
#' @return TRUE if x is a Money object
#' @export
is_money <- function(x) {
  inherits(x, "Money")
}

#' Format Money object.
#'
#' @param x Money object
#' @param ... Additional arguments (ignored)
#' @return Formatted string
#' @export
format.Money <- function(x, ...) {
  format_money(x)
}

#' Print Money object.
#'
#' @param x Money object
#' @param ... Additional arguments (ignored)
#' @export
print.Money <- function(x, ...) {
  cat(format_money(x), "\n")
  invisible(x)
}

#' Format a Money object as a string.
#'
#' @param money_obj Money object
#' @param symbol Currency symbol to use (NULL for code)
#' @param symbol_position "before" or "after" (default "before")
#' @return Formatted money string
#' @export
format_money <- function(money_obj, symbol = NULL, symbol_position = "before") {
  if (is.null(money_obj) || !is_money(money_obj)) {
    return(NA_character_)
  }

  # Convert minor units to major units
  major_value <- money_obj$minor_units / (10 ^ money_obj$decimals)

  # Format with proper decimal places
  formatted_value <- format(
    round(major_value, money_obj$decimals),
    nsmall = money_obj$decimals,
    big.mark = ",",
    scientific = FALSE
  )

  currency_indicator <- if (is.null(symbol)) money_obj$currency else symbol

  if (symbol_position == "after") {
    paste0(formatted_value, " ", currency_indicator)
  } else {
    paste0(currency_indicator, " ", formatted_value)
  }
}

#' Get amount in major units.
#'
#' @param money_obj Money object
#' @return Numeric amount in major units or NA
#' @export
money_amount <- function(money_obj) {
  if (is.null(money_obj) || !is_money(money_obj)) {
    return(NA_real_)
  }

  money_obj$minor_units / (10 ^ money_obj$decimals)
}

#' Get amount in minor units.
#'
#' @param money_obj Money object
#' @return Integer amount in minor units or NA
#' @export
money_minor_units <- function(money_obj) {
  if (is.null(money_obj) || !is_money(money_obj)) {
    return(NA_integer_)
  }

  money_obj$minor_units
}

#' Get currency code.
#'
#' @param money_obj Money object
#' @return Currency code or NA
#' @export
money_currency <- function(money_obj) {
  if (is.null(money_obj) || !is_money(money_obj)) {
    return(NA_character_)
  }

  money_obj$currency
}

#' Add two Money objects.
#'
#' @param a First Money object
#' @param b Second Money object
#' @return Money object or NULL if currencies differ
#' @export
money_add <- function(a, b) {
  if (!is_money(a) || !is_money(b)) {
    return(NULL)
  }

  if (a$currency != b$currency) {
    stop("Cannot add money with different currencies")
  }

  money_from_minor(a$minor_units + b$minor_units, a$currency)
}

#' Subtract two Money objects.
#'
#' @param a First Money object
#' @param b Second Money object
#' @return Money object or NULL if currencies differ
#' @export
money_subtract <- function(a, b) {
  if (!is_money(a) || !is_money(b)) {
    return(NULL)
  }

  if (a$currency != b$currency) {
    stop("Cannot subtract money with different currencies")
  }

  money_from_minor(a$minor_units - b$minor_units, a$currency)
}

#' Multiply Money by a scalar.
#'
#' @param money_obj Money object
#' @param multiplier Numeric multiplier
#' @return Money object
#' @export
money_multiply <- function(money_obj, multiplier) {
  if (!is_money(money_obj) || is.na(multiplier)) {
    return(NULL)
  }

  new_minor <- round(money_obj$minor_units * multiplier)
  money_from_minor(as.integer(new_minor), money_obj$currency)
}

#' Divide Money by a scalar.
#'
#' @param money_obj Money object
#' @param divisor Numeric divisor
#' @return Money object or NULL on division by zero
#' @export
money_divide <- function(money_obj, divisor) {
  if (!is_money(money_obj) || is.na(divisor) || divisor == 0) {
    return(NULL)
  }

  new_minor <- round(money_obj$minor_units / divisor)
  money_from_minor(as.integer(new_minor), money_obj$currency)
}

#' Addition operator for Money.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return Money object
#' @export
`+.Money` <- function(e1, e2) {
  if (is.numeric(e2)) {
    # Adding a scalar (in major units)
    money_add(e1, money(e2, e1$currency))
  } else {
    money_add(e1, e2)
  }
}

#' Subtraction operator for Money.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return Money object
#' @export
`-.Money` <- function(e1, e2) {
  if (is.numeric(e2)) {
    money_subtract(e1, money(e2, e1$currency))
  } else {
    money_subtract(e1, e2)
  }
}

#' Multiplication operator for Money.
#'
#' @param e1 Money object
#' @param e2 Numeric multiplier
#' @return Money object
#' @export
`*.Money` <- function(e1, e2) {
  if (is_money(e1) && is.numeric(e2)) {
    money_multiply(e1, e2)
  } else if (is.numeric(e1) && is_money(e2)) {
    money_multiply(e2, e1)
  } else {
    stop("Money can only be multiplied by a numeric value")
  }
}

#' Division operator for Money.
#'
#' @param e1 Money object
#' @param e2 Numeric divisor
#' @return Money object
#' @export
`/.Money` <- function(e1, e2) {
  if (!is.numeric(e2)) {
    stop("Money can only be divided by a numeric value")
  }
  money_divide(e1, e2)
}

#' Compare two Money objects for equality.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return TRUE if equal
#' @export
`==.Money` <- function(e1, e2) {
  if (!is_money(e1) || !is_money(e2)) {
    return(FALSE)
  }
  e1$currency == e2$currency && e1$minor_units == e2$minor_units
}

#' Compare Money for less than.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return TRUE if e1 < e2
#' @export
`<.Money` <- function(e1, e2) {
  if (!is_money(e1) || !is_money(e2)) {
    return(NA)
  }
  if (e1$currency != e2$currency) {
    stop("Cannot compare money with different currencies")
  }
  e1$minor_units < e2$minor_units
}

#' Compare Money for greater than.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return TRUE if e1 > e2
#' @export
`>.Money` <- function(e1, e2) {
  if (!is_money(e1) || !is_money(e2)) {
    return(NA)
  }
  if (e1$currency != e2$currency) {
    stop("Cannot compare money with different currencies")
  }
  e1$minor_units > e2$minor_units
}

#' Compare Money for less than or equal.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return TRUE if e1 <= e2
#' @export
`<=.Money` <- function(e1, e2) {
  e1 < e2 || e1 == e2
}

#' Compare Money for greater than or equal.
#'
#' @param e1 First Money object
#' @param e2 Second Money object
#' @return TRUE if e1 >= e2
#' @export
`>=.Money` <- function(e1, e2) {
  e1 > e2 || e1 == e2
}

#' Check if Money is negative.
#'
#' @param money_obj Money object
#' @return TRUE if negative
#' @export
money_is_negative <- function(money_obj) {
  if (!is_money(money_obj)) {
    return(NA)
  }
  money_obj$minor_units < 0
}

#' Check if Money is zero.
#'
#' @param money_obj Money object
#' @return TRUE if zero
#' @export
money_is_zero <- function(money_obj) {
  if (!is_money(money_obj)) {
    return(NA)
  }
  money_obj$minor_units == 0
}

#' Check if Money is positive.
#'
#' @param money_obj Money object
#' @return TRUE if positive
#' @export
money_is_positive <- function(money_obj) {
  if (!is_money(money_obj)) {
    return(NA)
  }
  money_obj$minor_units > 0
}

#' Get absolute value of Money.
#'
#' @param money_obj Money object
#' @return Money object with absolute value
#' @export
money_abs <- function(money_obj) {
  if (!is_money(money_obj)) {
    return(NULL)
  }
  money_from_minor(abs(money_obj$minor_units), money_obj$currency)
}

#' Negate Money.
#'
#' @param money_obj Money object
#' @return Negated Money object
#' @export
money_negate <- function(money_obj) {
  if (!is_money(money_obj)) {
    return(NULL)
  }
  money_from_minor(-money_obj$minor_units, money_obj$currency)
}

#' Allocate Money across multiple recipients.
#'
#' Uses largest remainder method to avoid losing pennies.
#'
#' @param money_obj Money object to allocate
#' @param ratios Numeric vector of allocation ratios
#' @return List of Money objects
#' @export
money_allocate <- function(money_obj, ratios) {
  if (!is_money(money_obj) || length(ratios) == 0) {
    return(NULL)
  }

  total_ratio <- sum(ratios)
  if (total_ratio == 0) {
    return(NULL)
  }

  total_minor <- money_obj$minor_units

  # Calculate base allocations
  allocations <- floor(ratios / total_ratio * total_minor)

  # Distribute remainder using largest remainder method
  remainder <- total_minor - sum(allocations)
  remainders <- (ratios / total_ratio * total_minor) - allocations

  while (remainder > 0) {
    max_idx <- which.max(remainders)
    allocations[max_idx] <- allocations[max_idx] + 1
    remainders[max_idx] <- -1  # Mark as used
    remainder <- remainder - 1
  }

  lapply(allocations, function(amt) {
    money_from_minor(as.integer(amt), money_obj$currency)
  })
}

#' Parse money from a string.
#'
#' @param money_string String like "USD 100.00" or "$100.00 USD"
#' @param default_currency Default currency if not specified
#' @return Money object or NULL on parse failure
#' @export
parse_money <- function(money_string, default_currency = "USD") {
  if (is.na(money_string) || is.null(money_string)) {
    return(NULL)
  }

  # Remove common currency symbols
  cleaned <- gsub("[$\u20ac\u00a3\u00a5]", "", money_string)

  # Remove commas and spaces used as thousand separators
  cleaned <- gsub(",", "", cleaned)
  cleaned <- trimws(cleaned)

  # Try to extract currency code and amount
  pattern <- "^([A-Z]{3})?\\s*([+-]?[0-9.]+)\\s*([A-Z]{3})?$"
  match_result <- regmatches(cleaned, regexec(pattern, cleaned, ignore.case = TRUE))[[1]]

  if (length(match_result) < 3) {
    return(NULL)
  }

  currency <- if (nchar(match_result[2]) > 0) {
    toupper(match_result[2])
  } else if (length(match_result) > 3 && nchar(match_result[4]) > 0) {
    toupper(match_result[4])
  } else {
    toupper(default_currency)
  }

  amount <- as.numeric(match_result[3])

  if (is.na(amount)) {
    return(NULL)
  }

  money(amount, currency)
}

File: bindings/ocaml/lib/safe_currency.ml
(* SPDX-License-Identifier: PMPL-1.0-or-later *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe currency handling with precise decimal arithmetic. *)

(** ISO 4217 currency codes. *)
type currency_code =
  | USD  (** United States Dollar *)
  | EUR  (** Euro *)
  | GBP  (** British Pound Sterling *)
  | JPY  (** Japanese Yen *)
  | CHF  (** Swiss Franc *)
  | CAD  (** Canadian Dollar *)
  | AUD  (** Australian Dollar *)
  | CNY  (** Chinese Yuan *)
  | INR  (** Indian Rupee *)
  | BRL  (** Brazilian Real *)
  | MXN  (** Mexican Peso *)
  | KRW  (** South Korean Won *)
  | SGD  (** Singapore Dollar *)
  | HKD  (** Hong Kong Dollar *)
  | NOK  (** Norwegian Krone *)
  | SEK  (** Swedish Krona *)
  | DKK  (** Danish Krone *)
  | NZD  (** New Zealand Dollar *)
  | ZAR  (** South African Rand *)
  | RUB  (** Russian Ruble *)
  | PLN  (** Polish Zloty *)
  | THB  (** Thai Baht *)
  | Other of string  (** Other ISO 4217 code *)

(** Money representation with minor units (cents, pence, etc.). *)
type t = {
  amount: int64;      (** Amount in minor units *)
  currency: currency_code;
  decimals: int;      (** Number of decimal places (usually 2, 0 for JPY/KRW) *)
}

type error =
  | Currency_mismatch
  | Overflow
  | Invalid_amount
  | Parse_error
  | Division_by_zero
  | Negative_amount

(** Get the number of decimal places for a currency. *)
let decimals_for_currency = function
  | JPY | KRW -> 0
  | _ -> 2

(** Get the multiplier for converting to minor units. *)
let multiplier decimals =
  let rec pow10 n acc =
    if n <= 0 then acc else pow10 (n - 1) (Int64.mul acc 10L)
  in
  pow10 decimals 1L

(** Convert currency code to ISO 4217 string. *)
let currency_to_string = function
  | USD -> "USD" | EUR -> "EUR" | GBP -> "GBP" | JPY -> "JPY"
  | CHF -> "CHF" | CAD -> "CAD" | AUD -> "AUD" | CNY -> "CNY"
  | INR -> "INR" | BRL -> "BRL" | MXN -> "MXN" | KRW -> "KRW"
  | SGD -> "SGD" | HKD -> "HKD" | NOK -> "NOK" | SEK -> "SEK"
  | DKK -> "DKK" | NZD -> "NZD" | ZAR -> "ZAR" | RUB -> "RUB"
  | PLN -> "PLN" | THB -> "THB" | Other s -> s

(** Parse currency code from ISO 4217 string. *)
let currency_of_string s =
  match String.uppercase_ascii s with
  | "USD" -> USD | "EUR" -> EUR | "GBP" -> GBP | "JPY" -> JPY
  | "CHF" -> CHF | "CAD" -> CAD | "AUD" -> AUD | "CNY" -> CNY
  | "INR" -> INR | "BRL" -> BRL | "MXN" -> MXN | "KRW" -> KRW
  | "SGD" -> SGD | "HKD" -> HKD | "NOK" -> NOK | "SEK" -> SEK
  | "DKK" -> DKK | "NZD" -> NZD | "ZAR" -> ZAR | "RUB" -> RUB
  | "PLN" -> PLN | "THB" -> THB | code -> Other code

(** Create money from minor units. *)
let of_minor_units amount currency =
  let decimals = decimals_for_currency currency in
  { amount; currency; decimals }

(** Create money from major units (e.g., dollars). *)
let of_major_units amount currency =
  let decimals = decimals_for_currency currency in
  let mult = multiplier decimals in
  { amount = Int64.mul (Int64.of_float amount) mult; currency; decimals }

(** Create money from string like "123.45". *)
let of_string str currency =
  let decimals = decimals_for_currency currency in
  try
    let str = String.trim str in
    (* Remove currency symbols if present *)
    let str =
      if String.length str > 0 && not (str.[0] >= '0' && str.[0] <= '9' || str.[0] = '-' || str.[0] = '+') then
        String.sub str 1 (String.length str - 1)
      else str
    in
    let str = String.trim str in
    match String.split_on_char '.' str with
    | [whole] ->
        let major = Int64.of_string whole in
        let mult = multiplier decimals in
        Ok { amount = Int64.mul major mult; currency; decimals }
    | [whole; frac] ->
        let major = Int64.of_string whole in
        let mult = multiplier decimals in
        let frac =
          if String.length frac > decimals then
            String.sub frac 0 decimals
          else
            frac ^ String.make (decimals - String.length frac) '0'
        in
        let minor = Int64.of_string frac in
        let is_negative = major < 0L || (major = 0L && String.length whole > 0 && whole.[0] = '-') in
        let amount =
          if is_negative then
            Int64.sub (Int64.mul major mult) minor
          else
            Int64.add (Int64.mul major mult) minor
        in
        Ok { amount; currency; decimals }
    | _ -> Error Parse_error
  with _ -> Error Parse_error

(** Format money to string. *)
let to_string money =
  let mult = multiplier money.decimals in
  let major = Int64.div money.amount mult in
  let minor = Int64.abs (Int64.rem money.amount mult) in
  if money.decimals = 0 then
    Printf.sprintf "%Ld" major
  else
    Printf.sprintf "%Ld.%0*Ld" major money.decimals minor

(** Format money with currency symbol. *)
let format ?(symbol=true) money =
  let amount_str = to_string money in
  if symbol then
    currency_to_string money.currency ^ " " ^ amount_str
  else
    amount_str

(** Add two money values. *)
let add a b =
  if a.currency <> b.currency then Error Currency_mismatch
  else
    let result = Int64.add a.amount b.amount in
    (* Check for overflow *)
    if (a.amount > 0L && b.amount > 0L && result < 0L) ||
       (a.amount < 0L && b.amount < 0L && result > 0L) then
      Error Overflow
    else
      Ok { a with amount = result }

(** Subtract two money values. *)
let sub a b =
  if a.currency <> b.currency then Error Currency_mismatch
  else
    let result = Int64.sub a.amount b.amount in
    (* Check for overflow *)
    if (a.amount > 0L && b.amount < 0L && result < 0L) ||
       (a.amount < 0L && b.amount > 0L && result > 0L) then
      Error Overflow
    else
      Ok { a with amount = result }

(** Multiply money by a scalar. *)
let mul money scalar =
  if scalar = 0L then Ok { money with amount = 0L }
  else
    let abs_amount = Int64.abs money.amount in
    let abs_scalar = Int64.abs scalar in
    (* Check for overflow before multiplying *)
    if abs_amount > Int64.div Int64.max_int abs_scalar then
      Error Overflow
    else
      Ok { money with amount = Int64.mul money.amount scalar }

(** Divide money by a scalar with rounding. *)
let div money divisor =
  if divisor = 0L then Error Division_by_zero
  else Ok { money with amount = Int64.div money.amount divisor }

(** Negate money value. *)
let negate money =
  if money.amount = Int64.min_int then Error Overflow
  else Ok { money with amount = Int64.neg money.amount }

(** Absolute value of money. *)
let abs money =
  if money.amount = Int64.min_int then Error Overflow
  else Ok { money with amount = Int64.abs money.amount }

(** Compare two money values. *)
let compare a b =
  if a.currency <> b.currency then
    Error Currency_mismatch
  else
    Ok (Int64.compare a.amount b.amount)

(** Check if money is zero. *)
let is_zero money = money.amount = 0L

(** Check if money is positive. *)
let is_positive money = money.amount > 0L

(** Check if money is negative. *)
let is_negative money = money.amount < 0L

(** Get amount in minor units. *)
let to_minor_units money = money.amount

(** Get amount as float (may lose precision). *)
let to_float money =
  let mult = multiplier money.decimals in
  Int64.to_float money.amount /. Int64.to_float mult

(** Create zero money in a currency. *)
let zero currency =
  of_minor_units 0L currency

(** Round to nearest minor unit after a calculation. *)
let round_half_up money =
  money  (* Already in minor units, no rounding needed *)

(** Split money into equal parts (handles remainders). *)
let split money n =
  if n <= 0 then Error Invalid_amount
  else
    let base_amount = Int64.div money.amount (Int64.of_int n) in
    let remainder = Int64.rem money.amount (Int64.of_int n) in
    let parts = Array.make n { money with amount = base_amount } in
    (* Distribute remainder to first parts *)
    for i = 0 to (Int64.to_int (Int64.abs remainder)) - 1 do
      let adjustment = if remainder > 0L then 1L else -1L in
      parts.(i) <- { parts.(i) with amount = Int64.add parts.(i).amount adjustment }
    done;
    Ok (Array.to_list parts)
